### PR TITLE
BUG: nunique not ignoring both None and np.nan

### LINKED ISSUE #37566: BUG: nunique not ignoring both None and np.nan
- [x] I have checked that this issue has not already been reported.

- [ ] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd
import numpy as np
pd.Series(['yes','yes', None, np.nan]).nunique(dropna = True)

```

#### Problem description
The output was 2.
It seems that nunique can only ignore None or np.nan but not both together. This seems odd to me and I think it 
should either be changed or the docs should reflect this.

#### Expected Output
I expected to get 1 as the result but I got 2.
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : db08276bc116c438d3fdee492026f8223584c477
python           : 3.7.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.49+
Version          : #1 SMP Sun Oct 18 19:43:35 PDT 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : C.UTF-8
LANG             : C.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.1.3
numpy            : 1.18.5
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.2.4
setuptools       : 46.1.3.post20200325
Cython           : 0.29.21
pytest           : 5.4.1
hypothesis       : 5.10.0
sphinx           : 3.0.2
blosc            : None
feather          : 0.4.1
xlsxwriter       : 1.2.8
lxml.etree       : 4.5.0
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.13.0
pandas_datareader: 0.9.0
bs4              : 4.9.0
bottleneck       : 1.3.2
fsspec           : 0.8.4
fastparquet      : None
gcsfs            : 0.6.1
matplotlib       : 3.2.1
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 3.0.3
pandas_gbq       : None
pyarrow          : 0.16.0
pytables         : None
pyxlsb           : None
s3fs             : 0.5.1
scipy            : 1.4.1
sqlalchemy       : 1.3.16
tables           : 3.6.1
tabulate         : 0.8.7
xarray           : 0.16.1
xlrd             : 1.2.0
xlwt             : 1.3.0
numba            : 0.48.0

</details>

### PR DESCRIPTION
**Pull request by GYHHAHA**
- [x] closes #37566
- [x] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/base.py [hunk 1]
```diff
@@ -1032,10 +1032,11 @@ def nunique(self, dropna: bool = True) -> int:
         >>> s.nunique()
         4
         """
-        uniqs = self.unique()
+        if dropna:
```
Comment (arw2019) [inline pandas/core/base.py:1035]: I'd do
``` python
obj = self.dropna() if dropna else self
return len(obj.unique())
```

File: pandas/tests/base/test_unique.py [hunk 2]
```diff
@@ -121,3 +121,11 @@ def test_unique_bad_unicode(idx_or_series_w_bad_unicode):
     else:
         expected = np.array(["\ud83d"], dtype=object)
         tm.assert_numpy_array_equal(result, expected)
+
+
+def test_nunique_dropna():
+    # test for #37566
```
Comment (arw2019) [inline pandas/tests/base/test_unique.py:127]: Do `# GH37566`

File: pandas/tests/base/test_unique.py [hunk 3]
```diff
@@ -121,3 +121,11 @@ def test_unique_bad_unicode(idx_or_series_w_bad_unicode):
     else:
         expected = np.array(["\ud83d"], dtype=object)
         tm.assert_numpy_array_equal(result, expected)
+
+
+def test_nunique_dropna():
```
Comment (arw2019) [inline pandas/tests/base/test_unique.py:126]: Parametrize on `dropna` (True/False)

File: pandas/tests/base/test_unique.py [hunk 4]
```diff
@@ -121,3 +121,11 @@ def test_unique_bad_unicode(idx_or_series_w_bad_unicode):
...
+    # test for #37566
+    s = pd.Series(['yes', 'yes', pd.NA, np.nan, None, pd.NaT])
+    res = s.nunique(dropna=True)
```
Comment (arw2019) [inline pandas/tests/base/test_unique.py:129]: just do `assert res == 1`

File: doc/source/whatsnew/v1.2.0.rst [hunk 5]
```diff
@@ -562,6 +562,7 @@ Other
 - Bug in :meth:`Index.union` behaving differently depending on whether operand is a :class:`Index` or other list-like (:issue:`36384`)
 - Passing an array with 2 or more dimensions to the :class:`Series` constructor now raises the more specific ``ValueError``, from a bare ``Exception`` previously (:issue:`35744`)
 - Bug in ``accessor.DirNamesMixin``, where ``dir(obj)`` wouldn't show attributes defined on the instance (:issue:`37173`).
+- Bug in :meth:`nunique` with ``dropna = True`` returns wrong result when different kinds of NA-like values exist (:issue:`37566`)
```
Comment (arw2019) [inline doc/source/whatsnew/v1.2.0.rst:565]:
```
Bug in :meth:`Series.nunique` with ``dropna = True`` was returning incorrect results when both ``NA`` and ``None`` missing values were present (:issue:`37566`)
```

**Review by arw2019 — CHANGES_REQUESTED** [review]
Thanks @GYHHAHA for the PR!

Some comments

**Comment by GYHHAHA** [comment]
This fails. So we can't use `dropna` in `IndexOpsMixin` even if this is inherited by both index and series which have `dropna` method? @arw2019 
```
Performing static analysis using mypy
pandas/core/base.py:1035: error: "IndexOpsMixin" has no attribute "dropna"  [attr-defined]
```

File: pandas/tests/base/test_unique.py [hunk 6]
```diff
@@ -121,3 +121,11 @@ def test_unique_bad_unicode(idx_or_series_w_bad_unicode):
     else:
         expected = np.array(["\ud83d"], dtype=object)
         tm.assert_numpy_array_equal(result, expected)
+
+
+@pytest.mark.parametrize("dropna", [True, False])
+def test_nunique_dropna(dropna):
+    # GH37566
+    s = pd.Series(["yes", "yes", pd.NA, np.nan, None, pd.NaT])
+    res = s.nunique(dropna=True)
```
Comment (arw2019) [inline pandas/tests/base/test_unique.py:130]: `res = s.nunique(dropna)` here, no?

**Review by arw2019 — CHANGES_REQUESTED** [review]
One more comment, o/w lgtm

**Review by arw2019 — APPROVED** [review]

File: doc/source/whatsnew/v1.2.0.rst [hunk 7]
```diff
@@ -562,6 +562,7 @@ Other
 - Bug in :meth:`Index.union` behaving differently depending on whether operand is a :class:`Index` or other list-like (:issue:`36384`)
 - Passing an array with 2 or more dimensions to the :class:`Series` constructor now raises the more specific ``ValueError``, from a bare ``Exception`` previously (:issue:`35744`)
 - Bug in ``accessor.DirNamesMixin``, where ``dir(obj)`` wouldn't show attributes defined on the instance (:issue:`37173`).
+- Bug in :meth:`Series.nunique` with ``dropna = True`` was returning incorrect results when both ``NA`` and ``None`` missing values were present (:issue:`37566`)
```
Comment (phofl) [inline doc/source/whatsnew/v1.2.0.rst:565]: Could you delete spaces around ``=``?

Comment (GYHHAHA) [reply doc/source/whatsnew/v1.2.0.rst]: done

**Comment by jreback** [comment]
looks good, just @phofl comment i think is left, ping on green.

**Comment by GYHHAHA** [comment]
cc @jreback

**Comment by jreback** [comment]
thanks @GYHHAHA very nice

File: pandas/core/base.py [hunk 8]
```diff
@@ -1032,11 +1032,8 @@ def nunique(self, dropna: bool = True) -> int:
         >>> s.nunique()
         4
         """
-        uniqs = self.unique()
-        n = len(uniqs)
-        if dropna and isna(uniqs).any():
-            n -= 1
-        return n
+        obj = remove_na_arraylike(self) if dropna else self
```
Comment (jorisvandenbossche) [inline pandas/core/base.py:1035]: Late comment, but would it be more performant if we removed the NAs *after* calculating the uniques? (assume the uniques is typically a much smaller array, and not this does an extra scan and creates a copy of the full array)

**Pull request merged**